### PR TITLE
fix(overlays): assign incremental id to overlay host

### PR DESCRIPTION
**Pull request by sean-perkins**
Issue number: Internal

---------

 

 

## What is the current behavior?

The counter for incrementing the `id` and `z-index` of an overlay is incremented whenever the `connectedCallback` is fired for an overlay. 

When an overlay is presented and/or conditionally rendered, the overlay `id` can increment by `n+2` instead of `n+1`.

## What is the new behavior?

- Increments all overlay ids consistently
- Removes legacy `ion-modal-{id}` and `ion-popover-{id}` logic
- Adds unit tests for the id behavior
  - Tests are split up into separate files so that the counter is always starting from `0`
- Adds an integration test with the Angular test app to verify conditional rendering behavior 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by sean-perkins — COMMENTED** [review]
Let me know if anyone else feels like a fixture would be useful here for the repeat overlay tests.

**Review by liamdebeasi — APPROVED** [review]
Good to go after the two comments are addressed.

File: angular/test/base/e2e/src/modal.spec.ts [hunk 1]
```diff
@@ -122,3 +122,25 @@ describe('when in a modal', () => {
     cy.get('#inputWithFloatingLabel').should('not.have.class', 'item-has-value');
   });
 });
+
+it('conditionally rendering a modal should increment the modal id', () => {
+  cy.visit('/modal-inline');
+
+  cy.get('#toggle-modal').click();
+
+  cy.get('ion-modal.conditional-modal')
```
[the remark this answers is not in the document]
Comment (sean-perkins) [reply angular/test/base/e2e/src/modal.spec.ts:131]: I would be fine with removing it from the Angular e2e tests. There is not any additional logic/behavior in the framework integrations for this aspect of overlays.

File: core/src/utils/overlays.ts [hunk 2]
```diff
@@ -54,11 +55,15 @@ export const prepareOverlay = <T extends HTMLIonOverlayElement>(el: T) => {
   if (typeof document !== 'undefined') {
     connectListeners(document);
   }
-  const overlayIndex = lastId++;
+  const overlayIndex = lastOverlayIndex++;
   el.overlayIndex = overlayIndex;
+};
+
+export const setOverlayId = <T extends HTMLIonOverlayElement>(el: T) => {
```
[the remark this answers is not in the document]
Comment (sean-perkins) [reply core/src/utils/overlays.ts:85]: Addressed here: https://github.com/ionic-team/ionic-framework/pull/27278/commits/ee67bcf41e52fe8318ccd06f946016d0d267c191

**Pull request merged**